### ISSUE TITLE
One-time crash on startup, related to locales

### ISSUE THREAD

**Issue by jyrkive** [wesnoth/wesnoth#1759]
Wesnoth (1.13.8+dev) crashed on startup when I launched it under a debugger. I don't know how to reproduce the crash, or how it can occur.

IRC discussion about the crash:

> [20:34:24] <JyrkiVesterinen> Got a crash on startup, under a debugger.
> [20:34:25] <JyrkiVesterinen> https://gist.github.com/jyrkive/db98bd3d3bbc5b90e0866174d09526ec
> [20:35:29] <JyrkiVesterinen> Everything in the wesnoth_message_format object is 0xDDDDDDDD. In Microsoft's debug CRT, that value means freed heap memory.
> [20:35:50] <JyrkiVesterinen> In other words, the wesnoth_message_format object is used after being destroyed.
> [20:36:39] <JyrkiVesterinen> The string passed to wesnoth_message_format::convert:
> [20:36:41] <JyrkiVesterinen> Lato-Regular.ttf,DejaVuSans.ttf,Andagii.ttf,DroidSansJapanese.ttf,DroidSansFallbackFull.ttf,Junicode-Regular.ttf
> [20:40:15] * Greg-Boggs (~greg_bogg@173.240.241.83) has quit IRC (Remote host closed the connection)
> [20:40:41] * Greg-Boggs (~greg_bogg@173.240.241.83) joins #wesnoth-dev
> [20:41:03] * Greg-Boggs (~greg_bogg@173.240.241.83) has quit IRC (Remote host closed the connection)
> [20:41:13] * Greg-Boggs (~greg_bogg@173.240.241.83) joins #wesnoth-dev
> [20:58:42] * irker099 (~irker@uruz.ai0867.net) joins #wesnoth-dev
> [20:58:42] <irker099> wesnoth: Charles Dang wesnoth:accelerated_rendering 9a97aec18428 / src/ (image.cpp image.hpp): Added a texture cache https://github.com/wesnoth/wesnoth/commit/9a97aec184285b83a45dcc02cae6c64b553e9992
> [20:58:42] <irker099> wesnoth: Charles Dang wesnoth:accelerated_rendering 68d94088ab85 / src/sdl/ (texture.cpp texture.hpp): Texture: converted internal handling to use a shared_ptr https://github.com/wesnoth/wesnoth/commit/68d94088ab8573c52362091ddedb8fd345daa23d
> [21:04:33] <JyrkiVesterinen> I'm not going to investigate that crash further. I can't figure out how it happens, and I can't reproduce it at will either.
> [21:04:59] <vultraz_iOS> well, that string is the font order list
> [21:06:19] <vultraz_iOS> loaded at font/font_config.cpp:152
> [21:06:23] <JyrkiVesterinen> Yes, and the call stack also shows that it happened in font::load_font_config().
> [21:06:57] <JyrkiVesterinen> However, as far as I can tell, the wesnoth_message_format object should NEVER be destroyed.
> [21:07:19] <vultraz_iOS> mystery indeed
> [21:07:57] <JyrkiVesterinen> Also, Boost calls wesnoth_message_format::get() for the same object just a moment before the point where it crashed... which suggests that the main thread destroyed the object  during that small window of time.
> [21:08:24] <vultraz_iOS> if you can't figure out the cause just file a bug
> [21:08:28] <JyrkiVesterinen> (Note that the crash occurred in a worker thread.)

**Comment by gfgtdf** [comment]
googleling this says that it can be fixed by building boost as a shard libary (dll, in b2 link=shared) instead of linking all code staticially to wesnoth.exe Maybe you coudl try that?

**Comment by jyrkive** [comment]
Well, because I have no way to reproduce, I would have no way to verify if it fixes the crash. :(

**Comment by gfgtdf** [comment]
ok then just for the record i'll link the page that said using shared boost libaries might help: https://stackoverflow.com/questions/26990412/c-boost-crashes-while-using-locale

**Comment by jyrkive** [comment]
Having read your link, I can now say with confidence that it would **not** help.

`boost::locale::basic_message<char>::write()` calls `wesnoth_message_format::get()` right before the `convert()` call, and that call had *succeeded*. If the problem was that the facet isn't accessible from within Boost, then the game would have crashed already when `get()` was called.

**Comment by singalen** [comment]
Discussion from development Discord channel:
[19:53] sinda: Another day, another quest. Have exactly this problem on iOS: https://stackoverflow.com/questions/39809811/boost-locale-compare-throws-exc-bad-access
Strange, all the objects involved are non-null.
21:42] jyrkive: AFAIK, the problem is use-after-free. It's just that the object in question should never be freed.
[21:44] jyrkive: And it's not macOS specific. It has been seen on Windows (me) and Linux (Arcanister).

My stack is:

```
#0	0x0000000000000000 in 0x00000000 ()
#1	0x00000001014ed7f8 in boost::locale::collator<char>::compare(boost::locale::collator_base::level_type, char const*, char const*, char const*, char const*) const at /Users/vic/src/wesnoth/wesnoth/projectfiles/Xcode-iOS/battleforwesnoth/../lib/iOS_boost_1_60/libs/boost/include/boost/locale/collator.hpp:81
#2	0x00000001014ed788 in translation::icompare(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) at /Users/vic/src/wesnoth/wesnoth/src/gettext_boost.cpp:429
#3	0x00000001013830c4 in gui2::dialogs::campaign_selection::sort_campaigns(gui2::window&, gui2::dialogs::campaign_selection::CAMPAIGN_ORDER, bool)::$_1::operator()(std::__1::shared_ptr<ng::level> const&, std::__1::shared_ptr<ng::level> const&) const at /Users/vic/src/wesnoth/wesnoth/src/gui/dialogs/campaign_selection.cpp:149
```

**Comment by CelticMinstrel** [comment]
This sounds like the same crash that I've gotten for awhile on the Mac when opening the file dialog (for example, go to map editor and click Load Map).

**Comment by singalen** [comment]
Might be the case. iOS version has access to two `iconv`s: Apple's version and GNU version. Sadly, glib and friends doesn't build on Apple `iconv`.

**Comment by CelticMinstrel** [comment]
If it _is_ binary incompatibilities, it just means that the libraries @ancestral builds with need to be carefully tuned to require the correct version of the dependencies... which is easier said than done, mind you.

**Comment by singalen** [comment]
Oh, it is a nightmare to just SOMEHOW build glib with its dependencies for iOS.

**Comment by jyrkive** [comment]
The binary incompatibility issue is clearly unrelated to the bug this report is about (a rare crash that occurs on all platforms). Please create a separate bug report about it.

**Comment by singalen** [comment]
Done: #2094

**Comment by sevu** [comment]
_Should be fixed by_ b780da4
A dupplicate of this issue is #2013
